pkg/sentry/platform/ring0/lib_arm64_unsafe.go

```diff
@@ -0,0 +1,112 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build arm64
+
+package ring0
+
+import (
+	"reflect"
+	"syscall"
+	"unsafe"
+
+	"gvisor.dev/gvisor/pkg/usermem"
+)
+
+const (
+	instSize = unsafe.Sizeof(uint32(0))
+	// The exception-vector-table is required to be 11-bits aligned.
+	// And the size is 0x800.
+	// Please see the documentation as reference:
+	// https://developer.arm.com/docs/100933/0100/aarch64-exception-vector-table
+	//
+	// But, golang does not allow to set a function's address to a specific value.
+	// So, for gvisor, I defined the size of exception-vector-table as 4K,
+	// filled the 2nd 2K part with NOPs.
+	// So that, I can safely move the 1st 2K part into the address with 11-bits alignment.
+	vectorsSize     = 0x800
+	vectorsSafeSize = 0x1000
+)
+
+func unsafeSlice(addr uintptr, length int) (slice []uint32) {
+	hdr := (*reflect.SliceHeader)(unsafe.Pointer(&slice))
+	hdr.Data = addr
+	hdr.Len = length / int(instSize)
+	hdr.Cap = length / int(instSize)
+	return slice
+}
+
+func getPageWithReflect(p uintptr, length int) []byte {
+	return *(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
+		Data: p,
+		Len:  length,
+		Cap:  length,
+	}))
+}
+
+// Work around: move ring0.Vectors() into a specific address with 11-bits alignment.
+//
+// According to the design documentation of Arm64,
+// the start address of exception vector table should be 11-bits aligned.
+// Please see the code in linux kernel as reference: arch/arm64/kernel/entry.S
+// But, we can't align a function's start address to a specific address by using golang.
+// We have raised this question in golang community:
+// https://groups.google.com/forum/m/#!topic/golang-dev/RPj90l5x86I
+// This function will be removed when golang supports this feature.
+//
+// There are 2 jobs were implemented in this function:
+// 1, move the start address of exception vector table into the specific address.
+// 2, modify the offset of each instruction.
+func rewriteVectors() {
+	vectorsBegin := reflect.ValueOf(Vectors).Pointer()
+	vectorsTable := unsafeSlice(vectorsBegin, vectorsSafeSize) // Now a []uint32
+	vectorsSize32 := vectorsSize / int(instSize)
+
+	offset := vectorsBegin & (1<<11 - 1)
+	if offset != 0 {
+		offset = 1<<11 - offset
+	}
+
+	NewVectorsBegin := vectorsBegin + offset
+
+	// 1st page.
+	page1 := getPageWithReflect(NewVectorsBegin, usermem.PageSize)
+	if err := syscall.Mprotect(page1, syscall.PROT_READ|syscall.PROT_WRITE|syscall.PROT_EXEC); err != nil {
+		panic(err)
+	}
+	// 2nd page.
+	page2 := getPageWithReflect(NewVectorsBegin+uintptr(usermem.PageSize), usermem.PageSize)
+	if err := syscall.Mprotect(page2, syscall.PROT_READ|syscall.PROT_WRITE|syscall.PROT_EXEC); err != nil {
+		panic(err)
+	}
+
+	offset = offset / instSize // By index, not bytes.
+	// Move exception-vector-table into the specific address, should uses memmove here.
+	for i := 1; i <= vectorsSize32; i++ {
```

Comment:
Does this work? I’m a bit confused about vectorsSize32 vs offset. Is vectorsSize32 the size of the whole region or just the number of vectors?

Edit: I get it now. I think maybe the word “size” is overloaded in context. Maybe new names? vectorsRegionSize => size of raw region in bytes. vectorsCodeLen => length of the code region (in instructions). Up to you.

Reply:
Done.